### PR TITLE
Automatically equip required items when in a level without them

### PR DESCRIPTION
**Pull request by nwinter**
### Automatically equip required items when in a level without them
If https://github.com/codecombat/codecombat/pull/7696 helps solve an issue where we don't own the proper items when going into a level in the InventoryModal, this solves similar issues if we somehow find ourselves in a level without the proper items (like navigating there by direct level link, or some other buggy flow).

This checks for requiredProperties, requiredGear, and restrictedGear, and does some just-in-time adjustments to what we have equipped in our session's heroConfig.inventory to meet the requirements.

There are probably some edge cases that aren't handled perfectly, like unequipping stuff in restricted slots that the default level config doesn't equip anything for, or dealing with rings, but this should do a good-enough job at the common cases for a fallback defense system. More edge cases are handled in the InventoryModal logic, but that's overkill complexity for here.

Fixes ENG-538.

To test, should just play a few complex level and inventory setups, for home and students-in-classroom-with-items-enabled accounts, and make sure this logic doesn't unequip some valid equipment choice. (It will log any changes it makes in the console.)

Screenshot 2024-09-15 at 13 42 56

### Handle another edge case with junior -> codecombat hero replacement
If we start to try to play CodeCombat with a CodeCombat Junior hero, it's possible to stay using the Junior hero. This adds a replacement to use a corresponding CodeCombat hero to another place that needs it.

## Summary by CodeRabbit

- **New Features**
	- Enhanced hero selection logic based on game mode, ensuring players have appropriate heroes for each level.
	- Implemented automatic equipment of default items to meet level requirements, improving player readiness.

- **Bug Fixes**
	- Resolved issues with hero type mismatches and inventory requirements, leading to a smoother gameplay experience.

### CODE REVIEW

**Comment by smallst** [comment]
before:
image
after:
image
LGTM

File: app/models/Level.js [hunk 1]
```diff
@@ -153,7 +153,20 @@ module.exports = (Level = (function () {
             }
             levelThang.components = [] // We have stored the placeholder values, so we can inherit everything else.
             heroThangType = __guard__(session != null ? session.get('heroConfig') : undefined, x => x.thangType)
-            if (heroThangType) { levelThang.thangType = heroThangType }
+            if (heroThangType) {
+              let juniorHeroReplacement
+              if (this.get('product', true) === 'codecombat-junior') {
+                // If we got into a codecombat-junior level with a codecombat hero, pick an equivalent codecombat-junior hero to use instead
+                juniorHeroReplacement = ThangTypeConstants.juniorHeroReplacements[_.invert(ThangTypeConstants.heroes)[heroThangType]]
```
Comment (smallst) [inline app/models/Level.js:160]: not blocking but maybe we can create a constant as `_.invert(heros)` rather than call function every time. (it's stable right?)

**Review by smallst — COMMENTED** [review]
the logic basic LGTM. 
but seems like it doesn't make sure the items can pass the level ?  is it normal ? (like depends on fuzzy verifier to fix ) 
image

File: app/models/Level.js (hunk 1 above)
Comment (nwinter) [reply app/models/Level.js:160]: Good refactor for next time we touch this

**Comment by nwinter** [comment]
Right, this doesn't do anything extra with clamped properties and min/max health, which we only apply in classroom mode. So in home version, it would sometimes be the case that they would need more armor or a better weapon. But at least the APIs should work.
